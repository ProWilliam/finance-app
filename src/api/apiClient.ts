// Config and Types
import config from '../../app.config';
import { FetchOptions } from '../types/api/Api.types';

const API_URL = config.extra.apiUrl;

const apiClient = async (endpoint: string, options?: FetchOptions) => {
  const { method = 'GET', headers = {}, body } = options || {};

  const config: RequestInit = {
    method,
    headers: {
      'Content-Type': 'application/json',
      ...headers,
    },
    body: JSON.stringify(body),
  };

  return globalThis.fetch(`${API_URL}${endpoint}`, config)
    .then(response => {
      if (!response.ok) {
        throw new Error(`Error: ${response.statusText}`);
      }
      return response.json();
    })
    .catch(error => {
      throw error;
    });

};

export default apiClient;