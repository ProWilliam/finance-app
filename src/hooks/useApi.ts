import { useState, useEffect, useCallback } from 'react';

// Component
import apiClient from '../api/apiClient';

// Types
import { FetchOptions } from '../types/api/Api.types';

const useApi = (endpoint: string, options?: FetchOptions) => {
  const [data, setData] = useState<any>(null);
  const [error, setError] = useState<Error | null>(null);
  const [loading, setLoading] = useState<boolean>(true);

  const fetchData = useCallback(
    async () => {
      setLoading(true);
      try {
        const result = await apiClient(endpoint, options);
        setData(result);
      } catch (err) {
        setError(err as Error);
      } finally {
        setLoading(false);
      }
  }, [endpoint, options]);

  useEffect(() => {
    if (!options?.manual) {
      fetchData();
    }
  }, [fetchData, options]);

  return { data, error, loading, refetch: fetchData };
};

export default useApi;