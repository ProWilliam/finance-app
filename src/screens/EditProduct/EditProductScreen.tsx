import React, { useEffect } from 'react';
import {View, Text, ScrollView } from 'react-native';
import { useRoute } from '@react-navigation/native';

// Hook and components
import useForm from '../../hooks/useForm';
import useApi from '../../hooks/useApi';
import AddProduct from '../../components/AddProduct/AddProduct';
import ButtonSubmit from '../../components/Button/ButtonSubmit';

// Styles and config
import style from './EditProduct.styles';
import config from '../../../app.config';
import { FormState } from '../../types/hook/useForm.type'
import apiClient from '../../api/apiClient';

const EditProductScreen: React.FC = () => {  

  const routePath = useRoute();
  const {id} = routePath.params as {id: string}

  const { data, loading } = useApi(config.extra.productUrl + id);
  const { formData, handleInputChange, setForm } = useForm();

  const submit = async () => {
    await apiClient(config.extra.productUrl + id , {method: "PUT", body: formData})
  }

  useEffect(() => {
    if(data){
      setForm(data as FormState)
    }
  }, [data]);

  return(
    <View style={style.containerRegister}>
      <Text style={style.title}>
        Actualizar Producto 
      </Text>
      {
        loading
        ?
        <Text>
          Loading...
        </Text>
        :
        <ScrollView style={style.scroll}>
          <AddProduct title='ID' placeHolder='Numero de ID' keySelect={'id'} onChange={handleInputChange} type='numeric' value={formData.id} />
          <AddProduct title='Nombre' placeHolder='Tarjeta Crédito' keySelect={'name'} onChange={handleInputChange} type='default' value={formData.name}/>
          <AddProduct title='Descripción' placeHolder='Descripción' keySelect={'description'} onChange={handleInputChange} type='default' value={formData.description}/>
          <AddProduct title='Logo' placeHolder='Logo'  keySelect={'logo'} onChange={handleInputChange} type='default' value={formData.logo}/>
          <AddProduct title='Fecha Liberación' placeHolder='22/02/2023'  keySelect={'date_release'} onChange={handleInputChange} type='date' value={formData.date_release}/>
          <AddProduct title='Fecha Revisión' placeHolder='22/02/2024'  keySelect={'date_revision'} onChange={handleInputChange} type='date' value={formData.date_revision}/>
          <ButtonSubmit title='Actualizar' color='send' press={submit} navigationRoot={config.extra.rootInfo} id={id}/>
          <ButtonSubmit title='Cancelar' color='edit' navigationRoot={config.extra.rootInfo} id={id}/>
        </ScrollView>
      }
    </View>
  );
}


export default EditProductScreen;