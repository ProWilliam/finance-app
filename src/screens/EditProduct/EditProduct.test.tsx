import React from 'react';
import renderer, { act, ReactTestRenderer } from 'react-test-renderer';
import { Text } from 'react-native';
import { useNavigation, useRoute, StackActions } from '@react-navigation/native';

import EditProductScreen from './EditProductScreen';
import useApi from '../../hooks/useApi';
import apiClient from '../../api/apiClient';

// Mock the `useApi` hook
jest.mock('../../hooks/useApi');

jest.mock('../../api/apiClient');

jest.mock('@react-navigation/native', () => ({
  useRoute: jest.fn(),
  useNavigation: jest.fn(),
  StackActions: {
    replace: jest.fn()
  }
}));

describe('<EditProductScreen />', () => {

  let tree: ReactTestRenderer;

  afterEach(() => {
    jest.clearAllMocks;
  });

  it('should render the component correctly', async () => {
    
    (useApi as jest.Mock).mockReturnValue({ data: {}, loading: false });
    
    (useRoute as jest.Mock).mockReturnValue({
      params: { id: '1' }, 
    });

    (useNavigation as jest.Mock).mockReturnValue({
      navigate: jest.fn(), 
    });

    (StackActions.replace as jest.Mock).mockReturnValue({
      type: 'replace',
      routeName: 'Home',
    });

    await act(async () => {
      tree = renderer.create(<EditProductScreen />);
    });

    expect(tree).toMatchSnapshot();
  });

  it('should show "Loading..." while data is loading', async () => {
    // Simula que los datos aÃºn no se han cargado
    (useApi as jest.Mock).mockReturnValue({ data: null, loading: true });
  
    await act(async () => {
      tree = renderer.create(<EditProductScreen />);
    });
  
    const textItems = tree.root.findAllByType(Text);
    const loadingText = textItems[1].props.children;

    expect(loadingText).toBe('Loading...');
  });

  it('should call dispatch function correctly', async () => {

    (useApi as jest.Mock).mockReturnValue({ data: { id: '1', name: 'Producto' }, loading: false });
  
    await act(async () => {
      tree = renderer.create(<EditProductScreen />);
    });
  
    const buttonSubmit = tree.root.findByProps({ title: 'Actualizar' });
    await act(async () => {
      buttonSubmit.props.press();
    });
  
    expect(apiClient).toHaveBeenCalledWith('bp/products/1', { method: 'PUT', body: { id: '1', name: 'Producto'} });
  });

  it('should update form status when data is received', async () => {

    (useApi as jest.Mock).mockReturnValue({ data: { id: '1', name: 'Producto' }, loading: false });
  
    await act(async () => {
      tree = renderer.create(<EditProductScreen />);
    });
  
    const addProductInput = tree.root.findByProps({ keySelect: 'name' });
    expect(addProductInput.props.value).toBe('Producto');
  });
});
