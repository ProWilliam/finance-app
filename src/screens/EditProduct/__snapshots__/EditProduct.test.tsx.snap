// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditProductScreen /> should render the component correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#fefefe",
      "flex": 1,
      "paddingBottom": 20,
      "paddingLeft": 60,
      "width": "100%",
    }
  }
>
  <Text
    style={
      {
        "fontSize": 34,
      }
    }
  >
    Actualizar Producto
  </Text>
  <RCTScrollView
    style={
      {
        "width": "100%",
      }
    }
  >
    <View>
      <View
        style={
          {
            "width": "80%",
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 14,
              "marginVertical": 10,
            }
          }
        >
          ID
        </Text>
        <View
          style={
            {
              "marginBottom": 5,
            }
          }
        >
          <TextInput
            blurOnSubmit={true}
            keyboardType="numeric"
            onBlur={[Function]}
            onChangeText={[Function]}
            placeholder="Numero de ID"
            placeholderTextColor="#6a6a6a"
            style={
              [
                {
                  "borderRadius": 5,
                  "borderWidth": 1,
                  "fontSize": 24,
                  "height": 50,
                  "paddingLeft": 20,
                  "width": "100%",
                },
                {
                  "borderColor": "#eaeaea",
                },
              ]
            }
            testID="error-message"
          />
        </View>
      </View>
      <View
        style={
          {
            "width": "80%",
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 14,
              "marginVertical": 10,
            }
          }
        >
          Nombre
        </Text>
        <View
          style={
            {
              "marginBottom": 5,
            }
          }
        >
          <TextInput
            blurOnSubmit={true}
            keyboardType="default"
            onBlur={[Function]}
            onChangeText={[Function]}
            placeholder="Tarjeta Crédito"
            placeholderTextColor="#6a6a6a"
            style={
              [
                {
                  "borderRadius": 5,
                  "borderWidth": 1,
                  "fontSize": 24,
                  "height": 50,
                  "paddingLeft": 20,
                  "width": "100%",
                },
                {
                  "borderColor": "#eaeaea",
                },
              ]
            }
            testID="error-message"
          />
        </View>
      </View>
      <View
        style={
          {
            "width": "80%",
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 14,
              "marginVertical": 10,
            }
          }
        >
          Descripción
        </Text>
        <View
          style={
            {
              "marginBottom": 5,
            }
          }
        >
          <TextInput
            blurOnSubmit={true}
            keyboardType="default"
            onBlur={[Function]}
            onChangeText={[Function]}
            placeholder="Descripción"
            placeholderTextColor="#6a6a6a"
            style={
              [
                {
                  "borderRadius": 5,
                  "borderWidth": 1,
                  "fontSize": 24,
                  "height": 50,
                  "paddingLeft": 20,
                  "width": "100%",
                },
                {
                  "borderColor": "#eaeaea",
                },
              ]
            }
            testID="error-message"
          />
        </View>
      </View>
      <View
        style={
          {
            "width": "80%",
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 14,
              "marginVertical": 10,
            }
          }
        >
          Logo
        </Text>
        <View
          style={
            {
              "marginBottom": 5,
            }
          }
        >
          <TextInput
            blurOnSubmit={true}
            keyboardType="default"
            onBlur={[Function]}
            onChangeText={[Function]}
            placeholder="Logo"
            placeholderTextColor="#6a6a6a"
            style={
              [
                {
                  "borderRadius": 5,
                  "borderWidth": 1,
                  "fontSize": 24,
                  "height": 50,
                  "paddingLeft": 20,
                  "width": "100%",
                },
                {
                  "borderColor": "#eaeaea",
                },
              ]
            }
            testID="error-message"
          />
        </View>
      </View>
      <View
        style={
          {
            "width": "80%",
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 14,
              "marginVertical": 10,
            }
          }
        >
          Fecha Liberación
        </Text>
        <View
          style={
            {
              "marginBottom": 5,
            }
          }
        >
          <View>
            <View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#eaeaea",
                    "borderRadius": 5,
                    "color": "#FFFFFF",
                    "opacity": 1,
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#1E2126",
                      "fontSize": 24,
                      "padding": 10,
                    }
                  }
                >
                  22/02/2023
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "width": "80%",
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 14,
              "marginVertical": 10,
            }
          }
        >
          Fecha Revisión
        </Text>
        <View
          style={
            {
              "marginBottom": 5,
            }
          }
        >
          <View>
            <View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#eaeaea",
                    "borderRadius": 5,
                    "color": "#FFFFFF",
                    "opacity": 1,
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#1E2126",
                      "fontSize": 24,
                      "padding": 10,
                    }
                  }
                >
                  22/02/2024
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#FFDD00",
            "borderRadius": 10,
            "height": 60,
            "justifyContent": "center",
            "marginVertical": 10,
            "opacity": 1,
            "paddingHorizontal": 15,
            "paddingVertical": 10,
            "width": "80%",
          }
        }
      >
        <Text
          style={
            {
              "color": "#1E2126",
              "fontSize": 14,
              "fontWeight": "bold",
              "padding": 10,
            }
          }
        >
          Actualizar
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#eaeaea",
            "borderRadius": 10,
            "height": 60,
            "justifyContent": "center",
            "marginVertical": 10,
            "opacity": 1,
            "paddingHorizontal": 15,
            "paddingVertical": 10,
            "width": "80%",
          }
        }
      >
        <Text
          style={
            {
              "color": "#1E2126",
              "fontSize": 14,
              "fontWeight": "bold",
              "padding": 10,
            }
          }
        >
          Cancelar
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
